### PR TITLE
Standardise credentials API (#4223) (#4163)

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Part of #4223
Part of #4163

# Rationale for this change
 

In order to expose extension points that relate to authorization and credentials, it helps to have a standard abstraction for representing this.

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: object_store/src/client/mod.rs [hunk 1]
```diff
@@ -503,6 +506,90 @@ impl GetOptionsExt for RequestBuilder {
     }
 }
 
+/// Provides credentials for use when signing requests
+#[async_trait]
+pub trait CredentialProvider: std::fmt::Debug + Send + Sync {
```
Comment (tustvold) [inline object_store/src/client/mod.rs:511]: This is the interface that will be made public as part of #4163

**Comment by tustvold** [comment]
@roeap perhaps you might be able to give this one a look, I believe you were interested in #4223

File: object_store/src/azure/credential.rs [hunk 2]
```diff
@@ -338,13 +342,15 @@ impl ClientSecretOAuthProvider {
...
+impl TokenProvider for ClientSecretOAuthProvider {
+    type Credential = AzureCredential;
+
     /// Fetch a token
     async fn fetch_token(
         &self,
         client: &Client,
         retry: &RetryConfig,
-    ) -> Result<TemporaryToken<String>> {
+    ) -> crate::Result<TemporaryToken<Arc<AzureCredential>>> {
```
Comment (roeap) [inline object_store/src/azure/credential.rs:353]:
```suggestion
    ) -> crate::Result<TemporaryToken<Arc<Self::Credential>>> {
```

probably not worth the change, but I find this a little bit more explicit when reading the code.

**Review by roeap — APPROVED** [review]
Looking great.

In a follow-up I would like to look into if we can make the scopes / resources in the oAuth flows somehow configurable, without out crating too much chaos :).

File: object_store/src/azure/credential.rs (hunk 2 above)
Comment (tustvold) [reply object_store/src/azure/credential.rs:353]: I think I prefer being explicit about the type returned

**Pull request merged**